### PR TITLE
ENH: dataframe now supports partial string selection

### LINKED ISSUE #1265: DataFrame loc on string timestamp fails
When we index a timestamp indexed dataframe with a slice of times encoded as strings everything is great

    result = df.loc['2015-01': '2015-02']

However, when we use a single string things fail

    result = df.loc['2015-01']

Probably there is some custom logic in dask.dataframe that is doing actual value comparisons of the input string against the `divisions`.  It would be nice to use whatever logic Pandas uses in order to coerce this input to the proper timestamp.

cc @jreback any pointers?

### PR DESCRIPTION
**Pull request by sinhrks**
Closes #1265.

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -390,13 +390,38 @@ def _loc(self, ind):
...
+                stop = index._maybe_cast_slice_bound(ind.stop, 'right', kind)
+            else:
+                stop = ind.stop
+            return slice(start, stop)
+
+        elif isinstance(ind, pd.compat.string_types):
+            start = index._maybe_cast_slice_bound(ind, 'left', 'loc')
+            stop = index._maybe_cast_slice_bound(ind, 'right', 'loc')
+            return slice(start, stop)
+        return ind
```
Comment (mrocklin) [inline dask/dataframe/core.py]: This function doesn't appear to use `self`.  Perhaps it should be a bare function instead of a method?

Comment (mrocklin) [inline dask/dataframe/core.py]: Ah, I see, it uses `self._pd.index`.  Perhaps just send in the index type as an input?

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -1387,12 +1413,19 @@ def columns(self, columns):
     def __getitem__(self, key):
         name = 'getitem-%s' % tokenize(self, key)
         if np.isscalar(key):
+
+            if isinstance(self._pd.index, (pd.DatetimeIndex, pd.PeriodIndex)):
+                if key not in self._pd.columns:
+                    return self._loc(key)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Is this behavior only for datetime indices?

File: dask/dataframe/core.py (hunk 1 above)
Comment (sinhrks) [inline dask/dataframe/core.py]: OK. What do you think split index-related functions to separate file, ``indexing.py`` (same name as pandas)?

Comment (mrocklin) [inline dask/dataframe/core.py]: Fine by me.  I defer to your opinion here.  I think that you probably have the most experience here.

Comment (sinhrks) [inline dask/dataframe/core.py]: ok, updated.

File: dask/dataframe/core.py (hunk 2 above)
Comment (sinhrks) [inline dask/dataframe/core.py]: yes it is. we can't rely on latter logic to avoid inf recursion.

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -1393,12 +1395,20 @@ def columns(self, columns):
     def __getitem__(self, key):
         name = 'getitem-%s' % tokenize(self, key)
         if np.isscalar(key):
+
+            if isinstance(self._pd.index, (pd.DatetimeIndex, pd.PeriodIndex)):
+                if key not in self._pd.columns:
+                    print('xxx')
```
Comment (mrocklin) [inline dask/dataframe/core.py]: print

**Comment by mrocklin** [comment]
+1 other than extra print statement

**Comment by mrocklin** [comment]
Merging on green light from travis

**Comment by mrocklin** [comment]
Thanks for this @sinhrks

**Pull request merged**